feat: add amo instruction layout system for zaamo extension

Sukuna0007Abhi opened the pull request:
**Add AMO Instruction Layout System for Zaamo Extension** 

fixes #223 

This PR implements a clean layout-based system for generating Atomic Memory Operation (AMO) instructions for the Zaamo extension, combining approaches suggested by @ddrcode and @ThinkOpenly.

✨ What's New
Layout Templates: Created 9 ERB templates (one per AMO operation) that automatically generate both word and doubleword variants
Organized Structure: Added layouts/ folder for templates and instructions/ folder for generated files
Automated Generation: Updated Rakefile to generate all 18 AMO instruction files from the layout templates
Better Descriptions: Each instruction now has proper long names and bullet-point descriptions
Schema Fixes: Fixed Zaamo.yaml extension definition with correct schema path

🎯 Result
Instead of manually maintaining 18 separate instruction files, we now have:

9 maintainable layout templates
Automatic generation of all AMO variants (.w and .d)
Consistent formatting and descriptions across all instructions
Easy to add new AMO operations in the future

🧹 Cleanup

Also updated .gitignore to exclude the 270+ generated CSR test files that were cluttering the repo.

Ready for review! 🎉 @AFOliveira @ThinkOpenly @dhower-qc

ThinkOpenly commented:
> **Add AMO Instruction Layout System for Zaamo Extension**
> 
> fixes #223

Thanks for your efforts!

> Layout Templates: Created 9 ERB templates (one per AMO operation) that automatically generate both word and doubleword variants

Sounds good, except that #223 is for Byte and Halfword atomics.

> Organized Structure: Added layouts/ folder for templates and instructions/ folder for generated files

I understand why you thought to do this, but this is not how we've been doing it so far. :slightly_smiling_face: The file extensions are sufficient to distinguish between layouts and instruction YAML. Let's leave everything where it is for now.

> Automated Generation: Updated Rakefile to generate all 18 AMO instruction files from the layout templates

OK

> 🎯 Result Instead of manually maintaining 18 separate instruction files, we now have:
> 
> 9 maintainable layout templates Automatic generation of all AMO variants (.w and .d) 

This is nice. This is also a step towards #361. Take a peek at that and see if that's something you wish to also include in this PR. Basically, all of these AMO instructions have not only the "width" variants, but also the "acquire/release" variants. For example, `amoadd` ends up with:
- `amoadd.b`, `amoadd.b.aq`, `amoadd.b.rl`, `amoadd.b.aq.rl`
- `amoadd.h`, `amoadd.h.aq`, `amoadd.h.rl`, `amoadd.h.aq.rl`
- `amoadd.w`, `amoadd.w.aq`, `amoadd.w.rl`, `amoadd.w.aq.rl`
- `amoadd.d`, `amoadd.d.aq`, `amoadd.d.rl`, `amoadd.d.aq.rl`
:exclamation: 

> Also updated .gitignore to exclude the 270+ generated CSR test files that were cluttering the repo.

These are files that we want to keep. The general process for "layout" files is to commit both the layout file and the generated files.

Sukuna0007Abhi commented:
> > **Add AMO Instruction Layout System for Zaamo Extension**
> > fixes #223
> 
> Thanks for your efforts!
> Thanks 
> > Layout Templates: Created 9 ERB templates (one per AMO operation) that automatically generate both word and doubleword variants
> 
> Sounds good, except that #223 is for Byte and Halfword atomics.
> 
> > Organized Structure: Added layouts/ folder for templates and instructions/ folder for generated files
> 
> I understand why you thought to do this, but this is not how we've been doing it so far. 🙂 The file extensions are sufficient to distinguish between layouts and instruction YAML. Let's leave everything where it is for now.
> Yes sir I thought the both layout flies and the instruction YAML will be mixed up that's but ok sir I will change that by today and should I leave the layout folder as it is and only deleting instruction folder.

> > Automated Generation: Updated Rakefile to generate all 18 AMO instruction files from the layout templates
> 
> OK
> 
> > 🎯 Result Instead of manually maintaining 18 separate instruction files, we now have:
> > 9 maintainable layout templates Automatic generation of all AMO variants (.w and .d)
> 
> This is nice. This is also a step towards #361. Take a peek at that and see if that's something you wish to also include in this PR. Basically, all of these AMO instructions have not only the "width" variants, but also the "acquire/release" variants. For example, `amoadd` ends up with:
> 
> * `amoadd.b`, `amoadd.b.aq`, `amoadd.b.rl`, `amoadd.b.aq.rl`
> * `amoadd.h`, `amoadd.h.aq`, `amoadd.h.rl`, `amoadd.h.aq.rl`
> * `amoadd.w`, `amoadd.w.aq`, `amoadd.w.rl`, `amoadd.w.aq.rl`
> * `amoadd.d`, `amoadd.d.aq`, `amoadd.d.rl`, `amoadd.d.aq.rl`
>   ❗
> 
> > Also updated .gitignore to exclude the 270+ generated CSR test files that were cluttering the repo.

> These are files that we want to keep. The general process for "layout" files is to commit both the layout file and the generated files.
> Ok then I will change the gitignore then both files will be there

Sukuna0007Abhi commented:
> > **Add AMO Instruction Layout System for Zaamo Extension**
> > fixes #223
> 
> Thanks for your efforts!
> Thanks 
> > Layout Templates: Created 9 ERB templates (one per AMO operation) that automatically generate both word and doubleword variants
> 
> Sounds good, except that #223 is for Byte and Halfword atomics.
> 
> > Organized Structure: Added layouts/ folder for templates and instructions/ folder for generated files
> 
> I understand why you thought to do this, but this is not how we've been doing it so far. 🙂 The file extensions are sufficient to distinguish between layouts and instruction YAML. Let's leave everything where it is for now.
> Yes sir I thought the both layout flies and the instruction YAML will be mixed up that's but ok sir I will change that by today and should I leave the layout folder as it is and only deleting instruction folder.

> > Automated Generation: Updated Rakefile to generate all 18 AMO instruction files from the layout templates
> 
> OK
> 
> > 🎯 Result Instead of manually maintaining 18 separate instruction files, we now have:
> > 9 maintainable layout templates Automatic generation of all AMO variants (.w and .d)
> 
> This is nice. This is also a step towards #361. Take a peek at that and see if that's something you wish to also include in this PR. Basically, all of these AMO instructions have not only the "width" variants, but also the "acquire/release" variants. For example, `amoadd` ends up with:
> 
> * `amoadd.b`, `amoadd.b.aq`, `amoadd.b.rl`, `amoadd.b.aq.rl`
> * `amoadd.h`, `amoadd.h.aq`, `amoadd.h.rl`, `amoadd.h.aq.rl`
> * `amoadd.w`, `amoadd.w.aq`, `amoadd.w.rl`, `amoadd.w.aq.rl`
> * `amoadd.d`, `amoadd.d.aq`, `amoadd.d.rl`, `amoadd.d.aq.rl`
>   ❗
> 
> > Also updated .gitignore to exclude the 270+ generated CSR test files that were cluttering the repo.

> These are files that we want to keep. The general process for "layout" files is to commit both the layout file and the generated files.

> Ok then I will change the gitignore then both files will be there

Sukuna0007Abhi commented:
Hi @ThinkOpenly,

I've implemented the AMO layout system with your byte/halfword approach and issue #361's acquire/release variants. The 9 layout files now generate all 144 instruction combinations (base, .aq, .rl, .aq.rl) automatically, replacing manual YAML maintenance. 

And also did the gitignore part and rewrite the folder structure 

Your conditional descriptions for "least-significant X of register" vs "value of register" are working perfectly across all operations.

Ready for Re-review! 🚀

Sukuna0007Abhi commented:
And @ThinkOpenly @dhower-qc @AFOliveira from the issue #361 I implemented the solution by these steps: 
Issue #361 Implementation:

Added acquire/release memory ordering variants to AMO instructions:

Generate 4 variants per operation: base, .aq, .rl, .aq.rl
Fixed bit encoding (hardcoded aq/rl bits instead of variables)
Dynamic instruction naming and descriptions based on memory ordering flags
Complete 144 instruction coverage (9 ops × 4 sizes × 4 aq/rl variants)

jordancarlin commented:
> I've implemented the AMO layout system with your byte/halfword approach (it will now work for 4 variants b, h w, d) and issue #361's acquire/release variants. The 9 layout files now generate all 144 instruction combinations (base, .aq, .rl, .aq.rl) automatically, replacing manual YAML maintenance.

Minor but important correction: the four suffixes should be nothing, `.aq`, `.rl`, and `.aqrl` (not `.aq.rl`).

Sukuna0007Abhi commented:
Sure,updating by today @ThinkOpenly 
& Thanks for suggesting @jordancarlin

Sukuna0007Abhi commented:
hi sir @ThinkOpenly and @jordancarlin I updated that and now all suffixes now follow the correct convention: .aqrl (not .aq.rl) for combined aq/rl. Thanks for catching this! Ready for re-review. @ThinkOpenly @dhower-qc

ThinkOpenly commented:
@Sukuna0007Abhi, could you check why this PR has 272 file changes? I think there is a lot of stuff here that shouldn't be here.

Also, there are some files missing: when we add layout generation, we commit the generated files here as well. So, while I certainly expect the `.layout` files in `spec/std/isa/inst/Zaamo`, I also want to see the generated `.yaml` files:

> Complete 144 instruction coverage (9 ops × 4 sizes × 4 aq/rl variants)

Sukuna0007Abhi commented:
Hi @ThinkOpenly sorry for delay I actually got stuck on rebase that's why there are several commits but I can't able to skip it now pls if only skip 29,july commits those are repeated commits and other members commits, that's why.But now the you can check there are not anymore 260+ files only there are required files I did.And now I will generate the files and commit here.Thank you

Sukuna0007Abhi commented on spec/std/isa/ext/Zaamo.yaml:
```diff
@@ -1,12 +1,12 @@
 kind: extension
 name: Zaamo
-long_name: Load-acquire/Store-release atomic instructions
+long_name: Atomic Memory Operations
```
Ok @AFOliveira I will reset that long_name but I will state the schema error that should be change

Sukuna0007Abhi commented:
I updated those parts @AFOliveira thanks for suggesting. And yet ready for Re-review and possibility of merge. @ThinkOpenly @AFOliveira @dhower-qc

dhower-qc reviewed (changes requested):
You're on the right path. Looks like a few changes aren't correct, though

ThinkOpenly reviewed (changes requested):
This is fantastic, and very close to complete! Great work!

It looks like some unnecessary changes leaked in, and I have a few minor requests.

Sukuna0007Abhi commented on spec/std/isa/inst/Zaamo/amoaddN.layout:
Hi sir @ThinkOpenly I'm updating changes but I'm a bit confused about should I rename those layout files like this amoadd.SIZE.AQRL.layout or amoadd[op].SIZE.AQRL.layout

Sukuna0007Abhi replied:
Updated @ThinkOpenly

Sukuna0007Abhi commented on Rakefile:
```diff
@@ -557,3 +599,19 @@ task "RVA20": "#{$root}/gen/profile/pdf/RVA20ProfileRelease.pdf"
 task "RVA22": "#{$root}/gen/profile/pdf/RVA22ProfileRelease.pdf"
 task "RVA23": "#{$root}/gen/profile/pdf/RVA23ProfileRelease.pdf"
 task "RVB23": "#{$root}/gen/profile/pdf/RVB23ProfileRelease.pdf"
+task "MC100-32-CTP-HTML": "#{$root}/gen/proc_ctp/pdf/MC100-32-CTP.html"
```
removed the part @dhower-qc

Sukuna0007Abhi commented on spec/std/isa/inst/Zaamo/amoadd.d.yaml:
```diff
@@ -6,19 +8,18 @@
 kind: instruction
 name: amoadd.d
-long_name: Atomic fetch-and-add doubleword
+long_name: Atomic fetch-and-add word
```
updated sir @dhower-qc

Sukuna0007Abhi commented on Rakefile:
```diff
@@ -382,6 +414,16 @@ namespace :gen do
+          Rake::Task["#{$resolver.std_path}/inst/Zaamo/#{op}.#{size}#{suffix}.yaml"].invoke
+        end
+      end
+    end
```
You're right @ThinkOpenly I updated that.

Sukuna0007Abhi commented:
Hi sir @ThinkOpenly @dhower-qc all cl checks will pass now except that regress smoke that check will be passed when we use my this PR #954

Sukuna0007Abhi commented on Rakefile:
```diff
@@ -5,6 +5,9 @@ require "sorbet-runtime"
 T.bind(self, T.all(Rake::DSL, Object))
 extend T::Sig
 
+require 'pathname'
+require 'erb'
```
Thanks sir @dhower-qc

Sukuna0007Abhi commented on Rakefile:
```diff
@@ -16,6 +19,9 @@ require "etc"
 $root = Pathname.new(__dir__).realpath
 $lib = $root / "lib"
 
+# Add lib directory to load path
+$LOAD_PATH.unshift($lib) unless $LOAD_PATH.include?($lib)
```
Removing it

Sukuna0007Abhi commented on spec/std/isa/inst/Zaamo/amoadd.b.aq.yaml:
```diff
@@ -0,0 +1,135 @@
+  vs: always
+  vu: always
+operation(): |
+  if (implemented?(ExtensionName::A) && (CSR[misa].A == 1'b0)) {
```
Updated it sir but the issue is no source for misa issue giving another cl check failure regress gen appendix, so when my #954 PR maybe merged then this misa part issue will be solved.Then i think all cl check will also pass.

Sukuna0007Abhi commented on spec/std/isa/inst/Zaamo/amoadd.b.rl.yaml:
```diff
@@ -0,0 +1,135 @@
+  }
+
+  XReg virtual_address = X[xs1];
+  X[xd] = amo<8>(virtual_address, X[xs2][7:0], AmoOperation::Add, 1'b0, 1'b1, $encoding);
```
updated it

Sukuna0007Abhi commented on spec/std/isa/inst/Zaamo/amoadd.b.aqrl.yaml:
```diff
@@ -0,0 +1,135 @@
+  }
+
+  XReg virtual_address = X[xs1];
+  X[xd] = amo<8>(virtual_address, X[xs2][7:0], AmoOperation::Add, 1'b1, 1'b1, $encoding);
```
updated it sir

Sukuna0007Abhi commented:
Pls check @ThinkOpenly @dhower-qc

Sukuna0007Abhi commented:
Hello sir @ThinkOpenly I need some help, I am actually confused a bit for Zaamo it's d and w & for Zabha it's b and h right and I did aqrl,aq & rl variant for both, so should I arrange the yaml files on their exact folders because it all are currently in zaamo, I will arrange regenerate them the b/h yamls in Zabha and d/w yamls in Zaamo. 

And one more thing the layout files it's actually working for b,h,w,d so should I differentiate them same layout in Zaamo but only d/w and same layout in Zabha b/h.

ThinkOpenly commented:
> Hello sir @ThinkOpenly I need some help, I am actually confused a bit for Zaamo it's d and w & for Zabha it's b and h right

Very good point.

> and I did aqrl,aq & rl variant for both, so should I arrange the yaml files on their exact folders because it all are currently in zaamo, I will arrange regenerate them the b/h yamls in Zabha and d/w yamls in Zaamo.

That makes sense, yes.

> And one more thing the layout files it's actually working for b,h,w,d so should I differentiate them same layout in Zaamo but only d/w and same layout in Zabha but only b/h.

Ideally, we don't replicate code. Can you make it so there is still a single layout file for all, but you generate the respective files into the Zaamo and Zabha directories?

One thing I just noticed is that the "definedBy" attributes need to be different between the two sets: `definedBy: Zaamo` and `definedBy: Zabha` as appropriate.

This is really helpful! Thanks!

ThinkOpenly commented:
> in Zabha there is also amocas(b/h) files and i think those are connected with Zacas so should i create a aqrl layout for this ? & One more question the single layout files for all where to put means in which directory? Should I create a directory?

It's slightly complicated, and the UDB YAML files for `amocas.{b,h}` may be slightly inaccurate:
- Most of the `amo` instructions for {w,d,q} are defined by Zaamo.
- `amocas.{w,d,q}` are defined by Zacas. Zacas depends on Zaamo.
- Most of the `amo` instructions for {b,h} are defined by Zabha. Zabha depends on Zaamo.
- `amocas.{b,h}` are defined by Zabha if Zacas is supported.

In that last line is well-represented currently in UDB, specifically `amocas.{b,h}` are only:
```
definedBy: Zabha
```
when they should be defined by "allOf" Zabha and Zacas. Do you want to fix that in this PR, or shall I open a new issue?

So, while it's not critical what directory these files appear in, it probably makes the most sense for:
- Zaamo-only instructions in Zaamo.
- Zacas-only instructions in Zacas.
- Zabha-only instructions in Zabha.
- and `amocas.{b,h}` in Zabha.

ThinkOpenly commented on spec/std/isa/inst/Zaamo/amoadd.d.aqrl.yaml:
```diff
@@ -0,0 +1,140 @@
+$schema: "inst_schema.json#"
+kind: instruction
+name: amoadd.d.aqrl
+long_name: Atomic fetch-and-add doubleword (acquire) (release) (acquire-release)
```
The "(acquire) (release) (acquire-release)" is a bit redundant. :-)
Could you put just a single parenthetical here, likely "(aquire-release)" in this case?

Sukuna0007Abhi replied:
ok sir updating it 👍

ThinkOpenly commented on spec/std/isa/inst/Zaamo/amoadd.d.aqrl.yaml:
```diff
@@ -0,0 +1,140 @@
+name: amoadd.d.aqrl
+long_name: Atomic fetch-and-add doubleword (acquire) (release) (acquire-release)
+description: |
+  Atomically with acquire ordering with release ordering:
```
This isn't what you intended, either, is it?

Sukuna0007Abhi replied:
I put the description as the same for it but ok sir i will change it a little bit

ThinkOpenly commented on spec/std/isa/inst/Zaamo/amoadd.SIZE.AQRL.layout:
```diff
@@ -0,0 +1,175 @@
+  Atomically<%= aq && rl ? " with acquire and release ordering" : (aq ? " with acquire ordering" : (rl ? " with release ordering" : "")) %>:
+
+    * Load the <%= current_size[:name] %> at address _xs1_
+    * Write the <%= %w[b h].include?(size) ? "sign-extended value" : (size == "w" ? "sign-extended value" : "loaded value") %> into _xd_
```
Why not combine "b", "h", and "w" here if they all emit the same string.
Does the behavior of "w" depend on XLEN, though, where it's only sign-extended if XLEN > 32?

Sukuna0007Abhi commented on spec/std/isa/inst/Zaamo/amoadd.SIZE.AQRL.layout:
```diff
@@ -0,0 +1,175 @@
+  vs: always
+  vu: always
+operation(): |
+  if (implemented?(ExtensionName::A) && MISA_CSR_IMPLEMENTED && (CSR[misa].A == 1'b0)) {
+    raise (ExceptionCode::IllegalInstruction, mode(), $encoding);
+  }
```
ok sir i got it now, thanks for pointing

Sukuna0007Abhi commented:
I updated all pls check sir @ThinkOpenly @dhower-qc @AFOliveira

ThinkOpenly commented:
The no-aq-no-rl cases for "d" and "w" are now missing, like `amoadd.d`. Maybe I'm jumping in while you are still working?

Sukuna0007Abhi commented on Rakefile:
```diff
@@ -382,6 +464,55 @@ namespace :gen do
+    #     { suffix: ".rl", aq: false, rl: true },
+    #     { suffix: ".aqrl", aq: true, rl: true }
+    #   ].each do |variant|
+    #     Rake::Task["#{$resolver.std_path}/inst/Zacas/amocas.#{size}#{variant[:suffix]}.yaml"].invoke
```
I understood and sorry for it sir, but after some more research and work I fixed it now automation will generate all just by updating the dependency, no more manual or duplication , thanks for pointing me the right way.

Sukuna0007Abhi replied:
Dependency Definition (Lines 162-192) & File Generation Rules (Lines 384-468) both will work by together perfectly

Sukuna0007Abhi commented:
pls check sir @ThinkOpenly sir @dhower-qc sir @AFOliveira

ThinkOpenly commented on Rakefile:
```diff
@@ -155,8 +158,38 @@ end
 
 
 namespace :test do
```
Maybe this isn't correct. Why is this located in the "test" namespace?

Sukuna0007Abhi replied:
Ok sir , I'm changing the approach

ThinkOpenly commented on Rakefile:
```diff
@@ -155,8 +158,38 @@ end
+
   desc "Check that instruction encodings in the DB are consistent and do not conflict"
-  task :inst_encodings do
+  task :inst_encodings => amo_instruction_files do
```
Is this necessary? Does this task not find the generated instructions in the configured/resolved architecture without this dependency?

Sukuna0007Abhi replied:
Yes sir at first it actually not finding those files but now I found a better one I am commiting it,

ThinkOpenly commented on Rakefile:
```diff
@@ -356,9 +389,88 @@ file "#{$resolver.std_path}/csr/Zicntr/mcountinhibit.yaml" => [
+        rl = variant[:rl]
+        erb = ERB.new(File.read($resolver.std_path / "inst/Zaamo/#{op}.SIZE.AQRL.layout"), trim_mode: "-")
+        erb.filename = "#{$resolver.std_path}/inst/Zaamo/#{op}.SIZE.AQRL.layout"
+        File.write(t.name, insert_warning(erb.result(binding), t.prerequisites.first))
```
Yeah. As a general rule, avoid changing generated files. :-)

ThinkOpenly commented on spec/std/isa/inst/Zaamo/amomaxu.w.yaml:
```diff
@@ -35,77 +33,74 @@ access:
   {
     if extension("A") then {
-      /* Get the address, X(rs1) (no offset).
+      /* Get the address, X(xs1) (no offset).
```
There shouldn't be any changes to any code in any of the `sail()` methods.

ThinkOpenly commented on Rakefile:
```diff
@@ -356,9 +359,88 @@ file "#{$resolver.std_path}/csr/Zicntr/mcountinhibit.yaml" => [
+  end
+end
+
+# Zacas variants (w, d, q) -> generated in Zacas directory using the same Zabha layout
+# ["w", "d", "q"].each do |size|
+#   # Define all acquire/release combinations
+#   aq_rl_variants = [
+#     { suffix: "", aq: false, rl: false },           # base instruction
+#     { suffix: ".aq", aq: true, rl: false },         # acquire only
+#     { suffix: ".rl", aq: false, rl: true },         # release only
+#     { suffix: ".aqrl", aq: true, rl: true }         # both acquire and release
+#   ]
+#
+#   aq_rl_variants.each do |variant|
+#     file "#{$resolver.std_path}/inst/Zacas/amocas.#{size}#{variant[:suffix]}.yaml" => [
+#       "#{$resolver.std_path}/inst/Zabha/amocas.SIZE.AQRL.layout",
+#       __FILE__
+#     ] do |t|
+#       aq = variant[:aq]
+#       rl = variant[:rl]
+#       erb = ERB.new(File.read($resolver.std_path / "inst/Zabha/amocas.SIZE.AQRL.layout"), trim_mode: "-")
+#       erb.filename = "#{$resolver.std_path}/inst/Zabha/amocas.SIZE.AQRL.layout"
+#       File.write(t.name, insert_warning(erb.result(binding), t.prerequisites.first))
+#     end
+#   end
+# end
```
If this code is no longer needed/used, remove it.

ThinkOpenly commented on spec/std/isa/inst/Zaamo/amomax.d.aqrl.yaml:
```diff
@@ -0,0 +1,142 @@
+$schema: "inst_schema.json#"
+kind: instruction
+name: amomax.d.aqrl
+long_name: Atomic MAX doubleword (acquire) (release) (acquire-release)
```
This probably isn't what you wanted. (It's not what I expected. ;-)

I expect something more like:
```suggestion
long_name: Atomic MAX doubleword (acquire-release)
```

ThinkOpenly commented on spec/std/isa/inst/Zaamo/amomax.d.aqrl.yaml:
```diff
@@ -0,0 +1,142 @@
+name: amomax.d.aqrl
+long_name: Atomic MAX doubleword (acquire) (release) (acquire-release)
+description: |
+  Atomically with acquire ordering with release ordering:
```
This doesn't read well, either. Perhaps:
```suggestion
  Atomically with acquire ordering and release ordering:
```
or better:
```suggestion
  Atomically with acquire and release ordering:
```

ThinkOpenly commented:
I think this only needs one more update to the "golden instruction appendix" to pass the CI tests. Are you willing to do so, @Sukuna0007Abhi ?
